src/NDItk/base/Record.hpp

```diff
@@ -0,0 +1,56 @@
+#ifndef NJOY_NDITK_BASE_RECORD
+#define NJOY_NDITK_BASE_RECORD
+
+// system includes
+#include <string>
+
+// other includes
+
+namespace njoy {
+namespace NDItk {
+namespace base {
+
+/**
+ *  @brief A base class for NDI records
+ */
+class Record {
+
+  /* fields */
+  std::string keyword_;
+
+protected:
+
+  /* constructor */
+
+  /**
+   *  @brief Constructor
+   *
+   *  @param[in] keyword   the keyword of the record
+   */
+  Record( std::string keyword ) : keyword_( std::move( keyword ) ) {}
+
+  /**
+   *  @brief Print the record keyword
+   *
+   *  @param[in] iter   the current position in the output
+   */
+  template< typename OutputIterator >
+  void print( OutputIterator& iter ) const {
+
+    for ( auto c : this->keyword() ) { *iter++ = c; }
```

Comment:
These iterators point to a range of characters (typically a string) so we can only do one character at a time. This is the way it was set up in ENDFtk so we continue to do it this way in ACEtk and now NDItk to keep consistency.

Reply:
The iterator typically points to a string, but we can define iterators to file streams too.

Reply:
This is pretty confusing to me also. The name of the function is print. I don't think see how that is a print. Perhaps a name change is in order.

There may be internal logic which makes this safe, but as an abstract thing (which it is as a base class), it strikes me as dangerous. Consider:
If record.keyword is a string of size n and iter point to a string, inString, of of size m, what are we doing. We're copying record.keyword into inString and then adding a '\n' after all copying is done, yes? What if m <= n? The null terminating char of inString is replaced by \n in the case of m=n, yes? for m<n we start accessing memory which is out of bounds

Reply:
I was not as clear as I should have been. If you use a normal iterator (such as iter = string.begin()), then yes: you have to ensure the range of characters has been initialised to a large enough size to be safe (which is also the case when you use standard library algorithms such as merge, copy, etc.). If this iterator is an insert_iterator (like back_inserter, ostream_iterator, etc.) then that takes care of growing the output range as we add characters.

Basically, we will use this like this:
```
std::string buffer;
auto output = std::back_inserter( buffer );
chunk.print( output );
```
where ```chunk``` is any type of NDItk component (a table, record, subrecord or a collection of records like the metadata). For the record: the buffer need not be a string per se, you can also use this to directly print to a file or a vector<char> if you wanted to.

ENDFtk and ACEtk use the same style of interface, hence us using it here again.

Reply:
One could use sfinae to do this, there's an actual std::output_iterator< T > in C++20 to test whether or not some type T is an output iterator. In this context however I don't think it matters. No user is going to directly use the print function.

Reply:
what concerns me is making this a long lasting program. In the next 50 years is any developer going to unwittingly pass in a standard iterator to this template and cause the compiler to write a really unsafe function? The compiler is not going to complain - it is perfectly valid to pass this in. The program may look like it works in certain situations. It would be really hard to debug this if a small accident occurred in development through developer ignorance or mistake.

I get where you're coming from and that these exist a lot - memcpy - for one. But this isn't a well known function. This exists just here in this program. Perhaps the name of the variable is enough here - OutputIterator - to document and note that this was only supposed to be an output iterator for safe use. But other than this implicit note, there is not much documenting how careful one must be with this template. It just makes me nervous to see this, you know?

Reply:
This style is very similar to the C++ standard library, so any seasoned C++ programmer should recognise this immediately (and understand the underlying limitations). Look here for an example: https://en.cppreference.com/w/cpp/algorithm/copy or https://en.cppreference.com/w/cpp/algorithm/transform

Now, on the reverse side of the argument I do understand. I have been guilty of forgetting to allocate a vector and just use the begin iterator to an empty vector in a STL algorithm. As a result, you always do things like this:
https://github.com/njoy/scion/blob/d828ef47aa0d60daece9cdd5ff486d76aa6c8882/src/scion/unionisation/unionise.hpp#L27 if you are not using something like a back_inserter.

Reply:
Alright, sounds good. If you are comfortable with this, I'm fine. Its good to have the discussion here. Hopefully, if an error does occur down the line, this conversation with hasten its resolution.